sdk/storage/azure-storage-blob/src/main/java/com/azure/android/storage/blob/transfer/SerialExecutor.java

```diff
@@ -0,0 +1,103 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package com.azure.android.storage.blob.transfer;
+
+import androidx.annotation.NonNull;
+import androidx.work.WorkManager;
+
+import java.util.ArrayDeque;
+import java.util.concurrent.Executor;
+
+/**
+ * Package private.
+ *
+ * An {@link Executor} that wrap and delegate work execution to another {@link Executor}.
+ * SerialExecutor ensure that the submitted works are executed serially using
+ * the wrapped executor.
+ *
+ * SerialExecutor is used by {@link TransferClient} for all its internal book keeping,
+ * which includes creating DB entries for transfer workers, querying DB for status,
+ * submitting transfer request to {@link WorkManager}.
+ *
+ * This is similar to a {@link android.os.Looper} that serially process messages from
+ * it's message queue. The difference is while Looper hold a dedicated thread throughout
+ * it lifetime, SerialExecutor delegate the work execution to another executor.
+ *
+ * @see <a href="https://developer.android.com/reference/java/util/concurrent/Executor">Executor</a>
+ */
+final class SerialExecutor implements Executor {
```

Comment:
How would you feel about just using `androidx.work.impl.utils.SerialExecutor` instead? It's public, but it's also in an `impl` package. However it does pretty much exactly the same thing as this class.

Reply:
The thing I'm worried about there is - using a type from their "impl" package which they can move around any time. Like we are allowed to do for public types in our "implementation|internal" namespace.

Reply:
That's true - in practice I haven't seen that happen, but it's definitely a consideration. They wouldn't have to do a major version bump if they wanted to break compatibility with this class, and we don't want to pin to an exact version since we're a library.

I guess my suggestion is probably to keep a copy of their version in our SDK and use the appropriate license + attribution, I see that your implementation and theirs are very similar but not quite identical.

Also, I know we do the same thing (put public classes not part of our public API in an `impl` package) - should we do the same for non-public classes? It feels like we should just so that all of that non-public-API code is in the same place.

Reply:
Also I just noticed that for the public impl classes or public methods that are not part of our public API we can add `@RestrictTo` which is an IDE hint to not use those classes / methods. We can use `LIBRARY` or `LIBRARY_GROUP` as appropriate.

Reply:
yes, I saw that androidx annotation used in jetpack libs. 

If we've a "customer consumable public type" on which we have any "public methods" that we don't want customers to use then we should use this annotation. So far for the types in "transfer namespace", we haven't come across any cases yet.

Reply:
reg `SerialExecutor`: I referred the "Serial" Executor example in here as well https://developer.android.com/reference/java/util/concurrent/Executor. In a way androidx SerialExecutor is also very close to this. 

Just wondering should we wait a bit more before we include one more TPL to list, I mean wait for the cases where we really cannot avoid copy-pasting 3rd party code. Like 310 time types.